build-logic/src/main/kotlin/ktlint-kotlin-common.gradle.kts

```diff
@@ -8,19 +8,32 @@ plugins {
 kotlin {
     // All modules, the CLI included, must have an explicit API
     explicitApi()
+    jvmToolchain(20)
 }
 
+val targetJavaVersion = JavaVersion.VERSION_1_8
+tasks.withType<JavaCompile>().configureEach {
+    options.release.set(targetJavaVersion.majorVersion.toInt())
+}
 tasks.withType<KotlinCompile>().configureEach {
-    compilerOptions {
-        jvmTarget.set(JvmTarget.JVM_1_8)
-    }
+    kotlinOptions.jvmTarget = targetJavaVersion.toString()
 }
 
-// compileJava task and compileKotlin task jvm target compatibility should be set to the same Java version.
-// For some reason, we fallback from toolchain using, see https://github.com/pinterest/ktlint/pull/1787
-tasks.withType<JavaCompile>().configureEach {
-    sourceCompatibility = JavaVersion.VERSION_1_8.toString()
-    targetCompatibility = JavaVersion.VERSION_1_8.toString()
+listOf(8, 11, 17).forEach { majorVersion ->
+    val jdkTest = tasks.register<Test>("testJdk$majorVersion") {
+        javaLauncher = javaToolchains.launcherFor {
+            languageVersion = JavaLanguageVersion.of(majorVersion)
+        }
+
+        description = "Runs the test suite on JDK $majorVersion"
+        group = LifecycleBasePlugin.VERIFICATION_GROUP
+
+        // Copy inputs from normal Test task.
+        val testTask = tasks.test.get()
+        classpath = testTask.classpath
+        testClassesDirs = testTask.testClassesDirs
+    }
+    tasks.named("check").configure { dependsOn(jdkTest) }
 }
```

Comment:
❗ 
This is the core change of this PR. Full explanation can be found on https://jakewharton.com/build-on-latest-java-test-through-lowest-java/
>No user is building these projects on different versions. No user is building these projects at all. Consumers are using the pre-built .jar which we ship to Maven Central built on a single version.

Instead of testing the library build with Java 8, the new flow is to build with Java 20 (link the publish workflow does) and test the bytecode generated by Java (with Java 8 target set) in listed Java runtimes. 

You can now run tests on jdk8 locally (`./gradlew testJdk8`)

Reply:
One side-effect is there was a comment JDK 8 tests are wonky on Windows. I haven't seen a details, I don't know why - but if this still is the case, I can ignore `testJdk8` task if running on windows machine

Reply:
There is a regression of toolchain using in #1787, see the reason in there.

Reply:
Can you summarize what was the regression specifically, it doesn't seem to be mentioned anywhere 👀

Reply:
Gotcha, thanks for letting me know that CI performance is a concern already 👍 
Given the issue was not about toolchains but rather parallelism and CI performance, I kept toolchains setup (to keep the build deterministic), but I introduced separate jobs which should behave somewhat the same way the current setup.
If this PR gets merged I can look into windows builds performance, for example, from my observations disabling VFS on windows saves ~30% of a CI build time.

Reply:
If you want to keep toolchain using, maybe we can do something like https://github.com/detekt/detekt/pull/5735.

Reply:
I wasn't sure which part you were referring to specifically (the link leads to PR not to specific comment or line of code), but I pushed another round of improvement. Feel free to review them if you want :) 

>If you want to keep toolchain using

I'd only clarify, it's not like I _want_, I don't know any other solution to make the build deterministic. Checking out `master` locally and running `./gradlew` throws:
```
* What went wrong:
Execution failed for task ':build-logic:compileKotlin'.
> Error while evaluating property 'compilerOptions.jvmTarget' of task ':build-logic:compileKotlin'.
   > Failed to calculate the value of property 'jvmTarget'.
      > Unknown Kotlin JVM target: 20
```
I'm not aware of any other solution that fixes the root cause of issue, so toolchains are not my _preference_ but the only solution I know, so they are somewhat _required_ 😉